File: internal/oauth/manager.go
package oauth

import (
	"context"
	"fmt"
	"sync"
	"time"
)

type TokenSource struct {
	mu     sync.Mutex
	store  *FileStore
	client *Client
	skew   time.Duration
}

func NewTokenSource(store *FileStore, client *Client) *TokenSource {
	return &TokenSource{
		store:  store,
		client: client,
		skew:   2 * time.Minute,
	}
}

func (s *TokenSource) ValidToken(ctx context.Context) (*TokenSet, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	token, err := s.store.Load(ctx)
	if err != nil {
		return nil, err
	}

	if time.Until(token.Expiry) > s.skew {
		return token, nil
	}
	if token.RefreshToken == "" {
		return nil, fmt.Errorf("access token is expiring and no refresh token is stored; run login again")
	}

	refreshed, err := s.client.Refresh(ctx, token.RefreshToken)
	if err != nil {
		return nil, fmt.Errorf("refresh token: %w", err)
	}

	// Some providers rotate refresh tokens; keep the previous one only when the
	// new response omits it.
	if refreshed.RefreshToken == "" {
		refreshed.RefreshToken = token.RefreshToken
	}
	if refreshed.AccountID == "" {
		refreshed.AccountID = token.AccountID
	}

	if err := s.store.Save(ctx, refreshed); err != nil {
		return nil, fmt.Errorf("persist refreshed token: %w", err)
	}

	return refreshed, nil
}

func (s *TokenSource) ForceRefresh(ctx context.Context) (*TokenSet, error) {
	s.mu.Lock()
	defer s.mu.Unlock()

	token, err := s.store.Load(ctx)
	if err != nil {
		return nil, err
	}
	if token.RefreshToken == "" {
		return nil, fmt.Errorf("no refresh token stored; run login again")
	}

	refreshed, err := s.client.Refresh(ctx, token.RefreshToken)
	if err != nil {
		return nil, fmt.Errorf("refresh token: %w", err)
	}
	if refreshed.RefreshToken == "" {
		refreshed.RefreshToken = token.RefreshToken
	}
	if refreshed.AccountID == "" {
		refreshed.AccountID = token.AccountID
	}

	if err := s.store.Save(ctx, refreshed); err != nil {
		return nil, fmt.Errorf("persist refreshed token: %w", err)
	}
	return refreshed, nil
}

File: pkce-codex-rs/src/oauth/manager.rs
use crate::oauth::{client::Client, store::FileStore};
use anyhow::{Result, anyhow};

#[derive(Debug, Clone)]
pub struct TokenSource {
    store: FileStore,
    client: Client,
    refresh_skew_secs: i64,
}

impl TokenSource {
    pub fn new(store: FileStore, client: Client) -> Self {
        Self {
            store,
            client,
            refresh_skew_secs: 120,
        }
    }

    pub fn with_refresh_skew_secs(mut self, refresh_skew_secs: i64) -> Self {
        self.refresh_skew_secs = refresh_skew_secs;
        self
    }

    pub async fn valid_token(&self) -> Result<crate::oauth::client::TokenSet> {
        let token = self.store.load()?;
        let now = chrono::Utc::now().timestamp();
        if token.expires_in - now > self.refresh_skew_secs {
            return Ok(token);
        }
        if token.refresh_token.is_empty() {
            return Err(anyhow!(
                "access token is expiring and no refresh token is stored; run login again"
            ));
        }

        let mut refreshed = self.client.refresh(&token.refresh_token).await?;
        if refreshed.refresh_token.is_empty() {
            refreshed.refresh_token = token.refresh_token;
        }
        if refreshed.account_id.is_none() {
            refreshed.account_id = token.account_id;
        }
        self.store.save(&refreshed)?;
        Ok(refreshed)
    }

    pub async fn force_refresh(&self) -> Result<crate::oauth::client::TokenSet> {
        let token = self.store.load()?;
        if token.refresh_token.is_empty() {
            return Err(anyhow!("no refresh token stored; run login again"));
        }

        let mut refreshed = self.client.refresh(&token.refresh_token).await?;
        if refreshed.refresh_token.is_empty() {
            refreshed.refresh_token = token.refresh_token;
        }
        if refreshed.account_id.is_none() {
            refreshed.account_id = token.account_id;
        }
        self.store.save(&refreshed)?;
        Ok(refreshed)
    }
}
